[InstCombine] Improve handling of `not` and free inversion.

goldsteinn opened the pull request:
- [InstCombine] Add tests for folding `(X + Y) - (W + Z)`; NFC
- [InstCombine] Add folds for `(X + Y) - (W + Z)`
- [InstCombine] Make `isFreeToInvert` check recursively.
- [InstCombine] Add additional tests for free inversion; NFC
- [InstCombine] add `getFreeInverted` to perform folds for free inversion of op
- [InstCombine] Replace `isFreeToInvert` + `CreateNot` with `getFreelyInverted`
- [InstCombine] Add transform for `~X + ~Y)` -> `-2 - Y - X`

goldsteinn committed:
[InstCombine] Make `isFreeToInvert` check recursively.

Some Instructions (select/min/max) are inverted by just inverting the
operands. So the answer of whether they are free to invert is really
just whether the operands are free to invert.

Differential Revision: https://reviews.llvm.org/D159056

goldsteinn committed:
[InstCombine] Add additional tests for free inversion; NFC

goldsteinn committed:
[InstCombine] add `getFreeInverted` to perform folds for free inversion of op

With the current logic of `if(isFreeToInvert(Op)) return Not(Op)` its
fairly easy to either 1) cause regressions or 2) infinite loops
if the folds we have for `Not(Op)` ever de-sync with the cases we
know are freely invertible.

This patch adds `getFreeInverted` which is able to build the free
inverted op along with check for free inversion to alleviate this
problem.

RKSimon commented:
I'm curious how you intend to commit this stack of PRs? Are you intending to squash + merge?

goldsteinn commented:
> I'm curious how you intend to commit this stack of PRs? Are you intending to squash + merge?

I was/am intending to push each commit discretely.
As they are approved I generally create patches, apply to main, then push.
Not the best workflow for PRs?
Is there a better way to create PR with many independent commits?

RKSimon commented:
I'm still trying to work out how best to create stacked PRs (especially when adding extra tests and want to show the diff) - everyone seems to have their own opinion on them right now :)

RKSimon commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -245,32 +246,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
     if (match(V, PatternMatch::m_AnyIntegralConstant()))
       return true;
 
+    if (Depth++ >= MaxAnalysisRecursionDepth)
```
(style) For recursion methods add Depth on the recursive calls themselves (isFreeToInvert(X,Y, Depth + 1)

goldsteinn replied:
This is something there is no consistency across LLVM with. I know in SelectionDAG we do `Depth; Depth + 1` but somewhere like ValueTracking we use `Depth++; Depth`.

Personally I like the `Depth++` way b.c you don't have to type `Depth + 1` everywhere (easier to not mess up imo), and it seems more consistent with what a see elsewhere in InstCombine / the middle end.

Unless you feel strongly would prefer to keep as is.

goldsteinn commented:
> I'm still trying to work out how best to create stacked PRs (especially when adding extra tests and want to show the diff) - everyone seems to have their own opinion on them right now :)

I don't really know either. If you know a better way than this I'm all ears.

goldsteinn commented:
> > I'm still trying to work out how best to create stacked PRs (especially when adding extra tests and want to show the diff) - everyone seems to have their own opinion on them right now :)
> 
> I don't really know either. If you know a better way than this I'm all ears.

What might improve the situation is if we could create branch on the llvm-project repo. I.e `users/<gh-username>/<branch-name>`. Then we could create series of PRs i.e
`main` <- `users/<user>/feature-patch-0` <- `users/<user>/feature-patch-1` <-... `users/<user>/feature-patch-N`

goldsteinn commented:
ping.

RKSimon commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -230,49 +230,114 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+      if (Builder != nullptr) {
+        A = getFreeInverted(A, A->hasOneUse(), Builder, Depth);
+        B = getFreeInverted(B, B->hasOneUse(), Builder, Depth);
+        assert(A != nullptr && B != nullptr);
```
(style) missing assert message

goldsteinn commented:
ping.

goldsteinn commented:
ping.

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -248,32 +249,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+
+    // Selects/min/max with invertible operands are freely invertible
+    if (match(V, m_Select(PatternMatch::m_Value(), PatternMatch::m_Value(A),
+                          PatternMatch::m_Value(B))) ||
```
With this generalization, the select case should check `shouldAvoidAbsorbingNotIntoSelect()` to avoid inverting logical and/or into non-canonical form.

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+  /// If Builder is nonnull, it will return a simplified ~V
+  /// If builder is null, it will return an arbitrary nonnull value (not
+  /// dereferenceable).
+  static Value *getFreeInverted(Value *V, bool WillInvertAllUses,
```
I'd call this getFreelyInverted().

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+  /// Return nonnull value if V is free to invert (with condition) regarding
+  /// WillInvertAllUses.
+  /// If Builder is nonnull, it will return a simplified ~V
+  /// If builder is null, it will return an arbitrary nonnull value (not
```
builder -> Builder for consistency with previous sentence.

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
-    if (match(V, m_Not(PatternMatch::m_Value())))
-      return true;
+    Value *A, *B;
+    if (match(V, m_Not(PatternMatch::m_Value(A))))
```
As you're touching all the code in here anyway, I'd add a `using namespace llvm::PatternMatch` inside this function, so make the implementation more idiomatic...

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+  /// dereferenceable).
+  static Value *getFreeInverted(Value *V, bool WillInvertAllUses,
+                                BuilderTy *Builder, unsigned Depth = 0) {
+    static Value *const kNonNull = reinterpret_cast<Value *>(uintptr_t(1));
```
`k` prefix is not typical LLVM style...

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+    }
+
+    // If `V` is of the form `Constant - A` then `-1 - V` can be folded into
+    // `A + (-1 - Constant)` if we are willing to invert all of the uses.
```
Comment is about sub, implementation is for ashr.

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+        match(V, m_Select(PatternMatch::m_Value(Cond), PatternMatch::m_Value(A),
+                          PatternMatch::m_Value(B))) &&
+        !match(V, PatternMatch::m_LogicalOp(PatternMatch::m_Value(),
+                                            PatternMatch::m_Value()));
```
Okay, I see you ended up basically doing the same thing as shouldAvoidAbsorbingNotIntoSelect() in a following commit. Would still be good to use the standard helper for that.

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,56 +233,116 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
+    if (IsSelect || match(V, m_MaxOrMin(PatternMatch::m_Value(A),
+                                        PatternMatch::m_Value(B)))) {
+      if (!getFreeInverted(A, A->hasOneUse(), /*Builder*/ nullptr, Depth) ||
+          !getFreeInverted(B, B->hasOneUse(), /*Builder*/ nullptr, Depth))
```
You could only check A without Builder upfront, and then check B with Builder, avoiding one duplicate check.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -4410,6 +4410,11 @@ Instruction *InstCombinerImpl::foldNot(BinaryOperator &I) {
+  // TODO: Could handle multi-use better by checking if all uses of NotOp (other
+  // than I) can be inverted.
+  if (Value *R = getFreeInverted(NotOp, NotOp->hasOneUse(), &Builder))
+    return replaceInstUsesWith(I, R);
```
There is a `sinkNotIntoXor()` fold that I think gets obsoleted by this one? Any test changes if you drop it?

goldsteinn replied:
Some tests still rely on it, but did't poke around that deeply.

goldsteinn replied:
Was actually just operand order changes. Dropping `sinkNotIntoXor`.

nikic commented:
Only looked at the first three commits, which imho could be a separate PR (maybe drop the rest from this one for now).

I think after adding getFreeInverted() it would be good to also go through existing users of isFreeToInvert() and replace them with getFreeInverted() where this is easy, e.g. in foldICmpOrXX(). Directly creating the inverted op instead of letting it get inverted separately does seem a lot more robust to me.

goldsteinn commented:
> Only looked at the first three commits, which imho could be a separate PR (maybe drop the rest from this one for now).
> 
> I think after adding getFreeInverted() it would be good to also go through existing users of isFreeToInvert() and replace them with getFreeInverted() where this is easy, e.g. in foldICmpOrXX(). Directly creating the inverted op instead of letting it get inverted separately does seem a lot more robust to me.

Thank you for the review.

Regarding replacing `getFreeInverted` with `isFreeToInvert` + (creating a `not`). So I do add a fold for using `getFreeInverted` in `not` fold. My feeling was we have places where we create a `not` in a canonicalizing effort and more places recognize `m_Not` than `isFreeToInvert`, so it made sense to let the new pattern be checked by other folds before stripping the `not`. But not married to that idea.

goldsteinn committed:
[InstCombine] Replace `isFreeToInvert` + `CreateNot` with `getFreelyInverted`

This is nearly an NFC, the only change is potentially to order that
values are created/names.

Otherwise it is a slight speed boost/simplification to avoid having to
go through the `getFreelyInverted` recursive logic twice to simplify
the extra `not` op.

goldsteinn commented:
> > Only looked at the first three commits, which imho could be a separate PR (maybe drop the rest from this one for now).
> > I think after adding getFreeInverted() it would be good to also go through existing users of isFreeToInvert() and replace them with getFreeInverted() where this is easy, e.g. in foldICmpOrXX(). Directly creating the inverted op instead of letting it get inverted separately does seem a lot more robust to me.
> 
> Thank you for the review.
> 
> Regarding replacing `getFreeInverted` with `isFreeToInvert` + (creating a `not`). So I do add a fold for using `getFreeInverted` in `not` fold. My feeling was we have places where we create a `not` in a canonicalizing effort and more places recognize `m_Not` than `isFreeToInvert`, so it made sense to let the new pattern be checked by other folds before stripping the `not`. But not married to that idea.

Okay, its split to 4 four commits related to the free inversion (re-titled).
The fourth commit is essentially NFC (although not quite due to change in inst order/names) and puts `getFreelyInverted` in a few places.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -4317,13 +4327,15 @@ Instruction *InstCombinerImpl::foldNot(BinaryOperator &I) {
+               "isFreeToInvert desynced with getFreelyInverted");
+        Value *InvMaxMin = Builder.CreateBinaryIntrinsic(InvID, NotX, NotY);
+        return replaceInstUsesWith(I, InvMaxMin);
+      }
```
I think this implements the same logic as getFreelyInverted() itself -- can we just call getFreelyInverted() on II?

(More generally, we should probably have a generic fold calling getFreelyInverted() here, which likely subsumes some of the other folds in this function. But maybe that's for a followup.)

goldsteinn replied:
Yup, dropping.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -2198,12 +2198,14 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
+      assert(NotOp0 != nullptr &&
+             "isFreeToInvert desynced with getFreelyInverted");
+      return BinaryOperator::CreateSub(NotOp1, NotOp0);
+    }
```
I strongly suspect that this fold is going to bite us now that we do the recursive walk. We can go back and forth between `(~X) - (~Y)` and `Y - X` as long as all of `X`, `Y`, `~X`, `~Y` are freely invertible.

Previously this would happen with constants (which this guards against), but I think now it can likely happen in more complex cases that root out at constants as well. I'm thinking something like `sub (select x, C1, C2), (select y, C3, C4)` might infinite loop, depending on whether the "binop of select" fold happens before this one or not. Or if not that, then maybe something like that with more operations in between.

Basically, for this transform it is important that we actually remove at least one instruction, not just "freely" invert in the sense that instruction count stays the same.

goldsteinn replied:
Ah, yeah thats a good point.

goldsteinn replied:
Adding tests for all binops we support with lhs/rhs pattern that is always freely invertible.

goldsteinn commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,58 +233,114 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
-      return isFreeToInvert(A, A->hasOneUse(), Depth) &&
-             isFreeToInvert(B, B->hasOneUse(), Depth);
+    if (IsSelect || match(V, m_MaxOrMin(m_Value(A), m_Value(B)))) {
+      if (!getFreelyInverted(A, A->hasOneUse(), /*Builder*/ nullptr, Depth))
```
Can do so elsewhere as well.

nikic commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -2106,6 +2106,99 @@ Instruction *InstCombinerImpl::visitGEPOfGEP(GetElementPtrInst &GEP,
 
+Value *InstCombiner::getFreelyInverted(Value *V, bool WillInvertAllUses,
+                                       BuilderTy *Builder, unsigned Depth) {
+  using namespace llvm::PatternMatch;
```
Can probably drop the `using namespace` now.

nikic commented on llvm/test/Transforms/LoopVectorize/ARM/mve-selectandorcost.ll:
```diff
@@ -86,7 +86,7 @@ define float @test(ptr nocapture readonly %pA, ptr nocapture readonly %pB, i32 %
 ; CHECK-NEXT:    [[DEC]] = add i32 [[BLOCKSIZE_ADDR_018]], -1
 ; CHECK-NEXT:    [[CMP_NOT:%.*]] = icmp eq i32 [[DEC]], 0
-; CHECK-NEXT:    br i1 [[CMP_NOT]], label [[WHILE_END]], label [[WHILE_BODY]], !llvm.loop [[LOOP2:![0-9]+]]
+; CHECK-NEXT:    br i1 [[CMP_NOT]], label [[WHILE_END]], label [[WHILE_BODY]], !llvm.loop [[LOOP3:![0-9]+]]
```
You can probably remove the diff in this test?

nikic commented on llvm/test/Transforms/InstCombine/free-inversion.ll:
```diff
@@ -0,0 +1,521 @@
+; CHECK-NEXT:    [[XX:%.*]] = xor i8 [[X:%.*]], 123
+; CHECK-NEXT:    [[TMP1:%.*]] = xor i8 [[Y:%.*]], -46
+; CHECK-NEXT:    [[YY_NEG:%.*]] = add i8 [[TMP1]], 1
+; CHECK-NEXT:    [[R:%.*]] = add i8 [[YY_NEG]], [[XX]]
```
Pre-existing issue: We're adding an extra instruction here. It looks like this bit in InstCombineNegator is responsible: https://github.com/llvm/llvm-project/blob/8748b3979ce414025825631dd77c27c58a4ed04b/llvm/lib/Transforms/InstCombine/InstCombineNegator.cpp#L445-L455 It's missing a IsTrulyNegation check. Will we run into an infinite loop here once we add it?

goldsteinn replied:
Yeah, you're right. Ill post patch for `IsTrulyNegatable` and add option for `MustConsumeInst` in `getFreelyInverted`.

goldsteinn replied:
See: https://github.com/llvm/llvm-project/pull/72767, will repost w/ `MustConsumeInst` flag once thats in.

goldsteinn replied:
Okay, update so that we ensure we save an inst.

goldsteinn committed:
[InstCombine] Add transform for `~X + ~Y)` -> `-2 - Y - X`

Proof: https://alive2.llvm.org/ce/z/36FySK

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -2205,12 +2227,23 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
-    return BinaryOperator::CreateSub(NotOp1, NotOp0);
+  {
+    auto m_AddSubInvertable = m_CombineOr(m_Add(m_Value(), m_ImmConstant()),
+                                          m_Sub(m_ImmConstant(), m_Value()));
```
What is this for?

goldsteinn replied:
The `sub` is unnecessary (will drop). But we rely on this fold for simplifying things like `(A + C0) - (B + C1)`

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -2205,12 +2223,22 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
+        (isFreeToInvert(Op0, Op0->hasOneUse(), ConsumesOp0) &&
+         isFreeToInvert(Op1, Op1->hasOneUse(), ConsumesOp1) &&
+         (ConsumesOp0 || ConsumesOp1) &&
+         !match(Op0, m_Sub(m_ImmConstant(), m_Value())))) {
```
Is this still needed?

goldsteinn replied:
Nope, dropping.

goldsteinn replied:
Done.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -2205,12 +2223,22 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
+    // otherwise this can inf loop.
+    bool ConsumesOp0, ConsumesOp1;
+    if ((match(Op0, m_Add(m_Value(), m_ImmConstant())) &&
+         match(Op1, m_Add(m_Value(), m_ImmConstant()))) ||
```
Isn't this going to do the wrong thing (or maybe assert?) if Op0 or Op1 are adds with extra uses?

Do I understand correctly that this folds `(X + C1) - (Y + C2)` into `(~C1 - X) - (~C2 - Y)` and then some other fold is going to fold that into `(X + Y) + (C1 - C2)`? If so, that seems super convoluted, should we just have an explicit fold to do that instead?

goldsteinn replied:
Yeah, thats probably cleaner. Will do.

goldsteinn replied:
Done.

goldsteinn committed:
[InstCombine] Add tests for folding `(X + Y) - (W + Z)`; NFC

goldsteinn committed:
[InstCombine] Add folds for `(X + Y) - (W + Z)`

If `Y` and `Z` are constant then we can simplify to `(X - W) + (Y -
Z)`. If `Y == Z` we can fold to `X - W`.

Note these transform exist outside of InstCombine. The purpose of this
commit is primarily to make it so that folds can generate these
simplifiable patterns without having to worry about creating an inf
loop.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -2224,12 +2224,17 @@ Instruction *InstCombinerImpl::visitSub(BinaryOperator &I) {
   // (~X) - (~Y) --> Y - X
   // This is placed after the other reassociations and explicitly excludes a
   // sub-of-sub pattern to avoid infinite looping.
```
The comment above is outdated / redundant with the one below.

nikic replied:
Well, at least the part about the "sub-of-sub pattern".

goldsteinn replied:
Although the whole thing is deprecated as ensuring we consume a `not` is what precludes inf loop.

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,66 +233,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
                                                 PatternMatch::m_Value()));
   }
 
+  /// Return nonnull value if V is free to invert (with condition) regarding
```
I'm not sure what the `(with condition)` bit here is supposed to mean.

nikic commented on llvm/include/llvm/Transforms/InstCombine/InstCombiner.h:
```diff
@@ -233,66 +233,38 @@ class LLVM_LIBRARY_VISIBILITY InstCombiner {
 
+  /// Return nonnull value if V is free to invert (with condition) regarding
+  /// WillInvertAllUses.
+  /// If Builder is nonnull, it will return a simplified ~V
```
nit: period at end of sentence.

nikic reviewed (approved):
LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).

goldsteinn commented:
> LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).

I'd be happy to improve the coverage. Why are you opposed to the commit as a whole?

nikic commented:
> > LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).
> 
> I'd be happy to improve the coverage. Why are you opposed to the commit as a whole? Also do you mean removed entirely, or is changing is to `ConsumeLHS && ConsumeRHS` okay?

It looks like a fold with very low value but very high risk. Limiting it to `ConsumeLHS && ConsumeRHS` (with `&&`, not `||`, and no special add/sub cases) would probably be acceptable.

goldsteinn commented:
> > > LGTM with the last commit removed (the test coverage is quite insufficient around m_AddSubInvertable, and I don't particularly want this fold anyway).
> > 
> > 
> > I'd be happy to improve the coverage. Why are you opposed to the commit as a whole? Also do you mean removed entirely, or is changing is to `ConsumeLHS && ConsumeRHS` okay?
> 
> It looks like a fold with very low value but very high risk. Limiting it to `ConsumeLHS && ConsumeRHS` (with `&&`, not `||`, and no special add/sub cases) would probably be acceptable.
Okay, done (added b.c noticed after changing the logic for `~Y - ~X` had a regression).

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -1662,6 +1662,24 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
+    // To ensure we can save instructions we need to ensure that we either
+    // consume both LHS/RHS (i.e they have a `not`).
+    bool ConsumesLHS, ConsumesRHS;
+    if (isFreeToInvert(LHS, LHS->hasOneUse(), ConsumesLHS) &&
+        isFreeToInvert(RHS, RHS->hasOneUse(), ConsumesRHS) && ConsumesLHS &&
+        ConsumesRHS) {
```
```suggestion
    if (isFreeToInvert(LHS, LHS->hasOneUse(), ConsumesLHS) && ConsumeLHS &&
        isFreeToInvert(RHS, RHS->hasOneUse(), ConsumesRHS) && ConsumesRHS) {
```
Can short-circuit the second call.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -1662,6 +1662,24 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
 
+  // (~X) + (~Y) --> (-2 - Y) - X
+  {
+    // To ensure we can save instructions we need to ensure that we either
```
```suggestion
    // To ensure we can save instructions we need to ensure that we
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:
```diff
@@ -1662,6 +1662,24 @@ Instruction *InstCombinerImpl::visitAdd(BinaryOperator &I) {
+             "isFreeToInvert desynced with getFreelyInverted");
+      Value *RHSMinus2 =
+          Builder.CreateSub(ConstantInt::get(RHS->getType(), -2), NotRHS);
+      return BinaryOperator::CreateSub(RHSMinus2, NotLHS);
```
I believe InstCombine's canonical form for this is `-2 - (X + Y)` rather than `(-2 - X) - Y`.

nikic reviewed (approved):
LGTM
